docs/api-spec.md

```diff
@@ -0,0 +1,324 @@
+
+## Resource Overview
+
+The Tekton Pipelines API provides a set of [Kubernetes Custom Resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) to manage run-to-completion workflows.
+
+### `TaskRun`
+
+A `TaskRun` represents an instantiation of a single execution of a `Task`. It can describe the steps of the Task directly.
+
+Its HTTP API endpoint is `/apis/tekton.dev/v1beta1/[parent]/taskruns`,  where  `[parent]` can refer to any string identifying a grouping of `TaskRun`s.
+
+For example, in the Kubernetes implementation `[parent]` refers to a Kubernetes namespace. Other implementations could interpret the string differently, including enabling hierarchies of resources containing `TaskRun`s, for example the API endpoint `/apis/tekton.dev/v1beta1/folders/my-folder/project/my-project/taskruns` could represent a hierarchy of "projects" that own `TaskRun`s.
+
+| HTTP Verb                 | Developer Access |
+|---------------------------|------------------|
+| Create (POST)             | REQUIRED         |
+| Patch (PATCH)*            | RECOMMENDED      |
+| Replace (PUT)             | RECOMMENDED      |
+| Delete (DELETE)           | OPTIONAL         |
+| Read (GET)                | REQUIRED         |
+| List (GET)                | REQUIRED         |
+| Watch (GET)               | OPTIONAL         |
+| DeleteCollection (DELETE) | OPTIONAL         |
+
+\* Kubernetes only allows JSON merge patch for CRD types. It is recommended that if allowed, at least JSON Merge patch be made available. [JSON Merge Patch Spec (RFC 7386)](https://tools.ietf.org/html/rfc7386)
+
+## Detailed Resources - v1beta1
+
+The following schema defines a set of REQUIRED or RECOMMENDED resource fields on the Tekton resource types. Whether a field is REQUIRED or RECOMMENDED is denoted in the "Requirement" column.
+
+Additional fields MAY be provided by particular implementations, however it is expected that most extension will be accomplished via the `metadata.labels` and `metadata.annotations` fields, as Tekton implementations MAY validate supplied resources against these fields and refuse resources which specify unknown fields.
+
+Tekton implementations MUST NOT require `spec` fields outside this implementation; to do so would break interoperability between such implementations and implementations which implement validation of field names.
+
+**NB:** All fields and resources not listed below are assumed to be **OPTIONAL**, not RECOMMENDED or REQUIRED. For example, at this time, support for `PipelineRun`s and for CRUD operations on `Task`s or `Pipeline`s is **OPTIONAL**.
```

Comment:
Why are `PipelineRuns` not included? They are part of the beta API and I see for reason for them to be excluded. 
I also have mixed feeling about having `Tasks` as OPTIONAL.
While it is possible to re-use Tasks (from the public or a private catalog) by injecting them on the fly in a `TaskRun`'s `taskSpec`, doing so would require a non-trivial amount of work for a Tekton use to move some of their workflows to such Tekton conformant platform.

Reply:
PipelineRuns are excluded purely because this is the minimal set I thought we could get away with. It's _definitely_ coming, but this PR is already huge :)